core/src/Torch/Raw/Tensor/Generic.hs

```diff
@@ -199,53 +205,176 @@ getDynamicDim
+numberFormat precision minSz xs = (format, scale, sz)
+  where
+    listParams :: (RealFloat a) => [a] -> (a, a, Bool, Bool, Bool)
+    listParams z =
+      f (NL.infinity, -NL.infinity, True, False, False) z
+      where
+        f z [] = z
+        f (minAcc, maxAcc, intModeAcc, hasInvalidAcc, hasNegInfAcc) (x:xs) = 
+            f acc xs
+          where
+            val = abs x
+            newHasNegInf = (x == -NL.infinity) || hasNegInfAcc
+            acc =
+              if (isNaN val || val == NL.infinity) then
+                (minAcc, maxAcc, intModeAcc, True, newHasNegInf) 
+              else 
+                (newMin, newMax, newIntMode, newHasInvalid, newHasNegInf)
+                  where
+                    !newHasInvalid = hasInvalidAcc
+                    !newMin = min minAcc val
+                    !newMax = max maxAcc val
+                    !newIntMode = if (intModeAcc) then
+                                    if (val == NL.infinity) then intModeAcc
+                                    else mod' val 1 == 0
+                                  else 
+                                    False
+    prec = precision
+    defaultScale = 1.0
+    (minAbs, maxAbs, intMode, hasInvalid, hasNegInf) = listParams xs
+    expMin = if (minAbs > 0.0) then 1 + (floor $ logBase 10 minAbs) else 1
+    expMax = if (maxAbs > 0.0) then 1 + (floor $ logBase 10 maxAbs) else 1
+    myMinSz = if hasInvalid then max minSz $ 3 + (if hasNegInf then 1 else 0)
+              else max minSz 2
+    (format, scale, sz) =
+      let expSz = (prec + 4 + (max (length (show expMax)) 
+                                   (length (show expMin)))) in
+      if (intMode) then
+        if (expMax > prec + 1) then
+          let tmpSz = max myMinSz expSz in
+          let format = "%" ++ show tmpSz ++ "." ++ show prec ++ "e" in
+          (format, defaultScale, tmpSz)
+        else
+          let tmpSz = max myMinSz (expMax + 1) in
+          let format = "%" ++ show tmpSz ++ ".0f" in
+          (format, defaultScale, tmpSz) 
+      else
+        if (expMax - expMin > prec) then
+          let tmpSz = max myMinSz expSz in
+          let format = "%" ++ show tmpSz ++ "." ++ show prec ++ "e" in
+          (format, 1.0, tmpSz) 
+        else
+          if (expMax > prec + 1) || (expMax < 0) then
+            let tmpSz = max myMinSz 7 in
+            let format = "%" ++ show tmpSz ++ "." ++ show prec ++ "f" in
+            (format, 10.0**fromIntegral(max minLogScale (expMax - 1)), tmpSz)
+          else
+            let tmpSz = if (expMax > 0) then expMax + 6 else 7 in
+            let format = "%" ++ show tmpSz ++ "." ++ show prec ++ "f" in
+            (format, 1.0, max myMinSz tmpSz)
+
+
+dispRawIntegral tensor = undefined
+
+dispRawRealFloat :: (THTensor t, RealFloat (HaskReal t)) => Ptr t -> IO ()
+dispRawRealFloat tensor
+ | length sz == 0 = putStrLn "Empty Tensor"
+ | length sz == 1 = do
+     putStrLn ""
+     let indexes = [ fromIntegral idx :: CLLong
```

Comment:
I'm almost thinking that we should split out these index traversal functions somewhere but I'm not sure how useful it would be since we delegate so much to TH. Something like `mapM1d`, `mapM2d`, etc, and then this function just becomes:

```haskell
dispRawRealFloat tensor =
  case whateverTheRankFunctionIs tensor of
    0 -> ...
    1 -> map1d fancyPrint tensor
    2 -> map2d fancyPrint tensor
    ...
```

You would preference `mapM` versions so that you can write a quick helper over `Identity` to extract a pure mapping function. Also, in an ideal world mapM1d, mapM2d, etc, are just using a generalized "construct the index traversal order" from the tensor dimensions and we can do away with the rank-specific helpers.